process: start coverage collection before bootstrap

joyeecheung opened the pull request:
This patch moves the dispatch of `Profiler.takePreciseCoverage`
to a point before the bootstrap scripts are run to ensure that
we can collect coverage data for all the scripts run after
the inspector agent is ready.

Before this patch `lib/internal/bootstrap/primordials.js` was not
covered by `make coverage`, after this patch it is.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

joyeecheung commented:
This also seems to fix the `async-hooks/test-callback-error` error in `make coverage` for me - probably because it no longer triggers the creation of an additional `JSBindingsConnection` (`AsyncWrap`) anymore.

CI: https://ci.nodejs.org/job/node-test-pull-request/20665/

addaleax commented on src/inspector_coverage.cc:
```diff
@@ -0,0 +1,172 @@
+        inspector->Connect(std::unique_ptr<V8CoverageSessionDelegate>(
+                               new V8CoverageSessionDelegate(this)),
+                           false);
+    session_.reset(session.release());
```
```suggestion
    session_ = std::move(session);
```

(assuming that works?)

addaleax commented on src/inspector_coverage.cc:
```diff
@@ -0,0 +1,172 @@
+    inspector::Agent* inspector = env->inspector_agent();
+    std::unique_ptr<inspector::InspectorSession> session =
+        inspector->Connect(std::unique_ptr<V8CoverageSessionDelegate>(
+                               new V8CoverageSessionDelegate(this)),
```
I think we are able to use `std::make_unique<…>(…)` now, if you want

addaleax commented on src/inspector_coverage.cc:
```diff
@@ -0,0 +1,172 @@
+
+  explicit V8CoverageConnection(Environment* env)
+      : BaseObject(env, env->coverage_connection()),
+        env_(env),
```
Is this field redundant? I think you could get the same thing through `this->env()`, right?

joyeecheung replied:
Yeah, I forgot that it's stored in BaseObject :)

joyeecheung commented:
Addressed reviews from @addaleax and fixed the quotes in the warning. Also I notice that sometimes undefined could be written to the file, so I've restored to an approach similar to the previous one and always overwrite the file in the callback once the callback is set.

CI: https://ci.nodejs.org/job/node-test-pull-request/20673/

targos commented on lib/internal/coverage-gen/with_profiler.js:
```diff
@@ -28,21 +23,14 @@ function writeCoverage() {
-    }));
-    const coverageInfo = JSON.parse(msg).result;
-    writeFileSync(target, JSON.stringify(coverageInfo));
+    internalBinding('coverage').end((msg) => {
```
this seems to be async now? should the try/catch be moved to the callback?

bcoe commented on src/inspector_coverage.cc:
```diff
@@ -0,0 +1,168 @@
+    explicit V8CoverageSessionDelegate(V8CoverageConnection* connection)
+        : connection_(connection) {}
+
+    void SendMessageToFrontend(
```
nice, seems like a huge benefit to move coverage standup and teardown into C++.

addaleax commented:
Looks like there are related failures in CI:

[Windows](https://ci.nodejs.org/job/node-test-binary-windows/23655/COMPILED_BY=vs2017,RUNNER=win10,RUN_SUBSET=1/testReport/junit/(root)/test/parallel_test_v8_coverage/):

```
AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:

134 !== 0

    at Object.<anonymous> (c:\workspace\node-test-binary-windows\test\parallel\test-v8-coverage.js:92:10)
```

[--without-intl](https://ci.nodejs.org/job/node-test-commit-linux-containered/10550/nodes=ubuntu1604_sharedlibs_withoutintl_x64/console):

```
/home/iojs/build/workspace/node-test-commit-linux-containered/out/Release/obj.target/node_lib/src/node_binding.o: In function `node::binding::RegisterBuiltinModules()':
node_binding.cc:(.text+0x14ae): undefined reference to `_register_coverage()'
collect2: error: ld returned 1 exit status
```

joyeecheung commented:
The failures on Windows:

```
C:\Users\joyee\projects\node>C:\Users\joyee\projects\node\Release\node.exe C:\Users\joyee\projects\node\test\parallel\test-v8-coverage.js
uv loop at [000001D549520AD0] has 2 active handles
[000001D54AF20CE0] async
        Close callback: 00007FF6CDFC2C40 <lambda_1e91a1e10c4ec4799db2c780f02b0099>::<lambda_invoker_cdecl> [c:\users\joyee\projects\node\src\node_platform.cc]:L275
        Data: 000001D54AEE2830
        (First field): 00019A740001A116
[000001D54AF21730] async
        Close callback: 00007FF6CDF4D1A0 node::inspector::`anonymous namespace'::DisposePairCallback [c:\users\joyee\projects\node\src\inspector\main_thread_interface.cc]:L99
        Data: 0000000000000000
Command Prompt - C:\Users\joyee\projects\node\Release\node.exe  C:\Users\joyee\projects\node\test\parallel\test-v8-coverage.js[7760]: c:\users\joyee\projects\node\src\debug_utils.cc:290: Assertion `0 && "uv_loop_close() while having open handles"' failed.
 1: 00007FF6CE0A723F node::DumpBacktrace+143 [c:\users\joyee\projects\node\src\debug_utils.cc]:L276
 2: 00007FF6CE047BD6 node::Abort+22 [c:\users\joyee\projects\node\src\node_errors.cc]:L166
 3: 00007FF6CE0481A3 node::Assert+131 [c:\users\joyee\projects\node\src\node_errors.cc]:L183
 4: 00007FF6CE0A6E28 node::CheckedUvLoopClose+216 [c:\users\joyee\projects\node\src\debug_utils.cc]:L290
 5: 00007FF6CDFA251E node::worker::Worker::`scalar deleting destructor'+126
 6: 00007FF6CE0A008D node::Environment::RunCleanup+701 [c:\users\joyee\projects\node\src\env.cc]:L524
 7: 00007FF6CE073836 node::Start+1334 [c:\users\joyee\projects\node\src\node.cc]:L831
 8: 00007FF6CE0731EE node::Start+382 [c:\users\joyee\projects\node\src\node.cc]:L877
 9: 00007FF6CE072D30 node::Start+1040 [c:\users\joyee\projects\node\src\node.cc]:L937
10: 00007FF6CDECC58C wmain+444 [c:\users\joyee\projects\node\src\node_main.cc]:L72
11: 00007FF6CEFE819C __scrt_common_main_seh+268 [d:\agent\_work\3\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl]:L288
12: 00007FF8B5643034 BaseThreadInitThunk+20
13: 00007FF8B7F13691 RtlUserThreadStart+33
```

I'll try looking into it later

bcoe commented:
quick update, haven't tested on any user-land modules yet; compiling the branch right now and will test tomorrow.

addaleax commented:
I haven’t tried it for this PR myself (yet), but I think #26138 and #26137 should take care of the Windows failures?

addaleax commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/20845/

bcoe commented:
@addaleax @joyeecheung I apologize for the slow turnaround, I've tested this change on a few user-land modules and it's working great 👍

as soon as tests pass, +1 from me.

joyeecheung commented:
Added `HAVE_INSPECTOR` guard for the `coverage` binding.

CI: https://ci.nodejs.org/job/node-test-pull-request/20856/

joyeecheung commented:
Landed in 2ae45d3
